### ISSUE TITLE
[CT-1673] [Feature] Dynamic impersonation of service accounts for one dbt run

### ISSUE THREAD

**Issue by alexandre-rassinoux** [dbt-labs/dbt-bigquery#440]
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt-bigquery functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

The current [impersonation configuration](https://docs.getdbt.com/reference/warehouse-setups/bigquery-setup#service-account-impersonation) within _profiles.yml_ takes affect only at connection time and applies to all queries that use that connection. It applies when getting credentials at client connection creation (see code [here](https://github.com/googleapis/python-bigquery/blob/eb49873176dee478617eb50472d44703abca53b5/google/cloud/bigquery/client.py#L170-L177)).
In a multiple dbt projects approach where a model in one project can have multiple childs in others, when we trigger a dbt run with the [plus operator (+)](https://docs.getdbt.com/reference/node-selection/graph-operators#the-plus-operator) in a select, it runs all models with the service account launched by the run, specified in the _profiles.yml_ of the dbt project. As a consequence, this service account must have a read/write access to all children which leads to a security issue because each dbt project is owned by different business teams. Each business team is identified by a different dbt service account in GCP, and must see only its scope in terms of data.

The idea is to be able to dynamically change the service account when running a model and its children located in different domains (with different permissions).

The proposed approach is described in the following diagram: the service account intended to run the model changes according to the dbt project in which the model is located.
![download](https://user-images.githubusercontent.com/64476570/208430087-e6fc9b67-d060-48f4-a2af-2fe0efd56603.png)

### Describe alternatives you've considered

Another alternative can be to run different dbt runs in a sequential approach, and stop using the plus operator (+) in the select.
Example:
I have A -> B -> C which are in three different dbt projects.
I run A with profiles_A, then run B with profiles_B, and C with profiles_C.

### Who will this benefit?

- Data Engineers which want to trigger only one DBT run with the plus operator in a secure way (by changing service account whether where you are located).
- Security teams which want to respect least privilege access by giving a service account only rights on its models, and not on children models.

### Are you interested in contributing this feature?

yes

### Anything else?

- Same concept as this issue, but for execution_project for billing purposes: https://github.com/dbt-labs/dbt-bigquery/issues/343
- If possible, it can be interesting to manage service account in a custom macro like custom database & schema: https://docs.getdbt.com/docs/build/custom-databases

**Comment by Fleid** [comment]
I agree with the problem, but not with solution. Not in its quality (stellar job writing it, thanks so much!) but in the direction.

To me we shouldn't allow multiple security context to coexist in a single dbt project. A dbt project is for a single team working on a single domain.
Instead we should facilitate collaboration across multiple dbt projects (teams, domains), each still representing a single security context.

Could you please have a look at what @jtcohen6 is doing over with [multi projects](https://github.com/dbt-labs/dbt-core/discussions/6725), and let me know if that would help you resolve your requirements?

**Comment by alexandre-rassinoux** [comment]
Hello Fleid,

First of all, I am sorry for taking so long to reply.

I agree that the approach I proposed is not optimal.
I have revised the approach by removing the cross-project invocation mechanism from the orchestrator. 
The multi-project feature you propose (which arrives soon) will provide even more flexibility by _importing_ only the models that I use in downstream dbt projects.

In terms of security, each service account will have read rights on upstream models, and read/write rights for its own models only.

Thank you for your help !
